State and Pair API changes, Freud RDFs

chrisjonesBSU opened the pull request:
This PR introduces some API changes to the `State` and `Pair` classes that I think will make it easier to manage all of the files involved.  Also, it uses Freud for RDF calculation rather than mdtraj.

Main API changes:

1. Creating a `State` object will automatically handle the creation of a `states` directory and a corresponding directory for each state.
2. The `add_state` function of `pair.py` now has a `calculate_target_rdf` option, so that the target RDF can be determined from the trajectory passed into the `State` object.  The user can still manually pass into `target_rdf` (either a text file or np array).
3. Because of No. 2 above, the `MSIBI` object that is created first gets passed into each `State` created. This is to ensure that the rdf related parameters are available when using `calculate_target_rdf=True` .  I couldn't really think of a better way to do this.
4. I added `calcualte_rdf.py` to the `utils` dir.  This has a function called `state_pair_rdf` using freud (via cmeutils). This is what is used when `calcualte_target_rdf=True` in `Pair.add_state()` This function could probably just be moved into `state.py`. I wasn't sure if there would be more involved that might warrant having an rdf related file in `utils`

Other changes
1. The hoomd system loaded in the `HOOMD2_HEADER` starts from frame 0 of the trajectory rather than the last frame.
2. The same rdf function above is used in the `compute_current_rdf()` function during iterations.

edit: keeping track of discussions/problems in https://etherpad.boisestate.edu/p/cme-msibi

Commit:
calculate rdf funciton, handle creating of state dirs

Commit:
Sate.target_rdf now a path to target_rdf.txt in state's dir

Commit:
fixing some errors in target rdf logic

Commit:
more robust checking of how to handle target_rdf

Commit:
create property func for target_rdf, rename func that calculates rdf

Commit:
update pair to use State attributes

Commit:
fix save parameters in save current rdf

Commit:
Merge branch 'master' of github.com:cmelab/msibi into feat/calc_rdf

Commit:
migrated target_rdf attribute to state.py

Commit:
trying a few things to access msibi context variables

Commit:
adding tutorials dir to gitignore

Commit:
remove manually created dirs in lj tutorial

Commit:
change handling of target rdf, no text files are saved

Commit:
move check # of rdf points from optimize.py to pair.add_state

Commit:
copy hoomd run template to state dirs

Commit:
using freud to calcualte current rdfs now

Commit:
fixes to find correct paths

Commit:
updated LJ tutorial with new api

Commit:
change func name of rdf calc

chrisjonesBSU commented:
Forgot to mention it, but the LJ tutorial was updated to show the revised api/workflow. I moved all of the hoomd run/rdf/detect equilibration to the bottom.

jennyfothergill commented on msibi/pair.py:
```diff
@@ -47,32 +48,59 @@ def __init__(self, type1, type2, potential, head_correction_form="linear"):
+                    )
+        if target_rdf:
+            if os.path.isfile(target_rdf):
+                try:
+                    target_rdf = np.loadtxt(target_rdf)
+                except Exception as e:
+                    print(e)
```
```suggestion
                target_rdf = np.loadtxt(target_rdf)
```
If we don't do anything with the error, no need for a try/except. Also just printing the error will let the program continue when we probably don't want it to. Also we probably don't want to catch the base exception.

chrisjonesBSU replied:
Good catch, fixed.

jennyfothergill commented on msibi/pair.py:
```diff
@@ -47,32 +48,59 @@ def __init__(self, type1, type2, potential, head_correction_form="linear"):
+            elif isinstance(target_rdf, np.ndarray):
+                pass
+
+        elif calculate_target_rdf:
+            target_rdf = state_pair_rdf(state, self)
```
We might want to make this logic a little more robust. Either the user provides an rdf or one is calculated. Unless is there another situation?
```suggestion
        else:
            target_rdf = state_pair_rdf(state, self)
```

chrisjonesBSU replied:
I guess in total there are 4 scenarios. Both are True, one or the other is given, or neither are given.  I added an additional else statement that catches the scenario where nothing is given to `target_rdf` and `calculate_target_rdf=False`

jennyfothergill commented on msibi/pair.py:
```diff
@@ -47,32 +48,59 @@ def __init__(self, type1, type2, potential, head_correction_form="linear"):
+        elif calculate_target_rdf:
+            target_rdf = state_pair_rdf(state, self)
+
+        if len(target_rdf) != state.opt.n_rdf_points:
+            raise ValueError(
+                    "The target RDF passed is not the same length as "
+                    "n_rdf_points set during the initialization of the "
+                    "MSIBI() class."
+                    )
```
maybe we should have this set automatically. It seems annoying to have to set that manually.

chrisjonesBSU replied:
What would be set automatically, `state.opt.n_rdf_points`? That comes from the `MSIBI` class which is created before State objects are created, so I don't see how it could be set automatically.  

This is kind of related to the second comment I left in #10.  If the user is going to create target RDFs separately, then how do we enforce that they were created with the same parameters that the query RDFs will be created? Like, same cutoff, number of x-axis bins, etc...

The parameters defined in the `MSIBI` class are automatically used when `calculate_target_rdf = True` and when calculating the query RDF.  So, in that sense this is handled automatically and the error wouldn't come up.  If the user is passing already generated RDFs, then they have to make sure the rdf related parameters passed to `MSIBI` match what was used when creating the target RDFs.

jennyfothergill replied:
I guess, do we want the user to ever pass generated RDFs? If so, we could just use len(rdf) for that value. And maybe we can remove n_rdf_points as an argument to MSIBI.

chrisjonesBSU replied:
If the RDFs are computationally expensive to generate, then I think there's value in giving the user an option to pass in already generated data for a target_rdf.

But somewhere the number of rdf points has to be defined, before `calculate_target_rdf=True` is called.  That can either be in `MSIBI` or defined each time a state is created, but that introduces potential issues that would have to be checked as well (it's the same value for all).

The only other thing I can think of is to leave the `n_rdf_points` as an optional parameter in the `MSIBI` class, and if the user is passing in an already generated `target_rdf` then its length is used to set the `MSIBI.n_rdf_points` attribute which is then used later when finding the query RDF.

chrisjonesBSU replied:
But then that doesn't address the other issues that come up with mismatching RDF parameters.  This would make sure that the bin sizes are the same, but what about cutoff?  Are both `n_rdf_points` and `rdf_cutoff` left as optional parameters that can be derived from whatever is passed into `target_rdf`?  Length of data and maximum x-axis value?  Do we make sure that these values are the same every time a `target_rdf` is passed to a state-pair combination?

Ultimately, I think it's ok to require the user to do some due diligence and understand the process enough to make sure they are passing in the correct info, like bin size and cutoff when initializing an optimization process.  Even if it means few extra checks are done. It's their responsibility to make sure that the same parameters they used when creating their own RDFs are passed into the `MSIBI` class. 

I like the idea conveyed by the doc strings in the `MSIBI class` **"Management class for orchestrating an MSIBI optimization"**

You set key info in this class that is used during the rest of the process.  If you are using `calcualte_target_rdf`, then you will know that everything matches.  If you aren't, then it's up to you to make sure you have everything in order so that things do match.  I think the best we can do is catch any potential issues and throw an error, not try to derive things from the RDFs passed in and set up the variable space for them.

jennyfothergill replied:
maybe a solution is to have both classes (state and pair) connected to the MSIBI object through some add_state/add_pair functions

jennyfothergill commented on msibi/utils/general.py:
```diff
@@ -4,7 +4,13 @@
 
+def get_msibi_instance():
+    for obj in gc.get_objects():
+            if isinstance(obj, msibi.optimize.MSIBI):
```
is the indentation weird here?

chrisjonesBSU replied:
Ooops, this was supposed to be removed.

jennyfothergill committed:
Merge branch 'master' of github.com:cmelab/msibi into feat/calc_rdf

jennyfothergill committed:
add cmeutils

jennyfothergill committed:
remove environment stuff from notebook

jennyfothergill reviewed (commented):
Nice work! I think this PR takes care of a lot of the overhead for an msibi user. 
I tried running the example in LJ_Tutorial, but it errored out when it didn't find the gsd files:
```
OSError: No such file: msibi/msibi/tutorials/lj/trajectory0.5.gsd
```
so let's add those.
A couple more suggestions:

- [x] move comment in notebook to markdown text (cleaner IMO)
- [x] LJ_Tutorial: move the commented out hoomd cells up and [hide them](https://jupyterbook.org/interactive/hiding.html) with some sort of explanation like: "if you want to see how these trajectories were created click here"
- [x] Drop mdanalysis as a dependency
- [ ] Add/update docstrings for any new/changed parameters
- [x] Update unit tests
- [ ] (last step before merging) Bump version in __version__ and environment.yml and push a tagged version.

Commit:
remove get instance func, not used

Commit:
Merge branch 'feat/calc_rdf' of github.com:chrisjonesBSU/msibi into feat/calc_rdf

Commit:
better handling or errors, addl. else statement for target rdf

Commit:
including needed gsd files for tutorial

jennyfothergill committed:
move hoomd run stuff

jennyfothergill committed:
just antefoyer

jennyfothergill committed:
hide hoomd cells, move imports, move comments to markdown

jennyfothergill committed:
steo -> step

jennyfothergill committed:
force add the target RDFs

jennyfothergill committed:
save output and all working

jennyfothergill committed:
update deps

Commit:
update doc strings, remove topology parameter and _is_gsd attribute

Commit:
update doc strings for Pair and Pair.add_state

Commit:
add to doc strings, added underscore to initialize function

chrisjonesBSU commented:
Considering the progress made in cmeutils [polymer tools](https://github.com/cmelab/cmeutils/pull/7) should we add the functionality for an MSIBI process to accept a bond length and bond angle distribution and create the necessary potentials for its simulations?

If so, where do you think this should live?

jennyfothergill commented:
> Considering the progress made in cmeutils [polymer tools](https://github.com/cmelab/cmeutils/pull/7) should we add the functionality for an MSIBI process to accept a bond length and bond angle distribution and create the necessary potentials for its simulations?
> 
> If so, where do you think this should live?

I think it would make sense for it to live in MSIBI. We have msibi/pair.py, maybe msibi/bonded.py? and have functions for taking a bonded/angle distribution to a potential.

jennyfothergill committed:
MSIBI class acts as container for pairs and states

jennyfothergill committed:
states is now private attribute, option of providing rdf removed,

rdf function now method of pair class

jennyfothergill committed:
opt now private

jennyfothergill committed:
update example to new schema

jennyfothergill committed:
delete (function moved to pair)

jennyfothergill committed:
remove python2 workarounds

jennyfothergill committed:
remove python2 workarounds

jennyfothergill committed:
remove select pairs and reformat

chrisjonesBSU committed:
Merge pull request #1 from jennyfothergill/feat/obj

Use MSIBI class as manager for pair and states

Commit:
changing max_frames to reasonable number, change start frame from read_gsd back to -1 after convo with Eric.

Commit:
merge upstream

Commit:
Merge branch 'master' of github.com:cmelab/msibi into feat/calc_rdf

Commit:
Merge branch 'feat/calc_rdf' of github.com:chrisjonesBSU/msibi into feat/calc_rdf

Commit:
remove some dcd stuff, take up if statements that check for hoomd2

Commit:
added Bonds and Angles class (bonds.py)

chrisjonesBSU commented:
There's a tutorial directory (`peek`) and notebook that shows how the workflow should work with passing in CG bond and angle parameters.

The values I  used for `k` is just a guess for debugging, the values for `r0` and `theta` are roughly based on values I've seen in bond and angle distributions.  

It seems to run without any errors, but some of the potentials are coming out inverted, but others aren't. Not sure what the issue could be.

Commit:
remove bit of code that used the cmeutils polymers PR to make the CG traj files

jennyfothergill commented:
Working with @chrisjonesBSU to try and debug:

- [x] inverted potentials are most likely due to the linear head correction. Do we want to force the linear head correction to have a negative slope? example:
![Screen Shot 2021-05-25 at 5 34 44 PM](https://user-images.githubusercontent.com/39961845/119581377-a912a600-bd7f-11eb-8446-f5565b06e292.png)

- [x] the "query.gsd" file created in the peek tutorial has no particle information -- what's going on?

- [x] after `MSIBI.optimize()` is finished running but the python kernel is still active (e.g. in a notebook), the spawned md_worker processes do not stop. Fix?

- [x] I think smoothing the target RDF might give improvement.

jennyfothergill committed:
add x values to rdf

jennyfothergill committed:
add context manager for pool

jennyfothergill committed:
oops fix indexing

jennyfothergill committed:
make plots easier to read

- set better y limits
- prevent head/tail corrections from being hidden

jennyfothergill committed:
try forcing the head correction slope to be negative

jennyfothergill committed:
set better ylims, smooth target rdf,

jennyfothergill committed:
plot potentials

Commit:
add pool.join and close; switch some formatting to f-strings

Commit:
add query traj attribute to state, and query flag to get_state_rdf

jennyfothergill committed:
remove multiprocessing/simplify

jennyfothergill committed:
only support gsd

jennyfothergill committed:
rework the hoomd_templating:

- move all the HOOMD script parts out of state.py and into hoomd_run_template.py,
- don't copy the template file -- instead import its contents and append it to the header,
- only initialize bonds/angles if they exist,
- add default=None for bonds/angles

jennyfothergill committed:
switch popen for run

jennyfothergill committed:
add rmin for rdf and potential to stop divide by zero error

Commit:
change exclude bonded default to False

Commit:
Merge branch 'feat/calc_rdf' of github.com:chrisjonesBSU/msibi into feat/calc_rdf

Commit:
change negative target rdf data points to zero

Commit:
change potential correction step backto adding the boltzmann inverse

Commit:
add momentum to dynamic kwarg when writing out gsd files

Commit:
add exclude_bonded as an optimizer param; set excluded_bonded and smooth from state._opt rather than explicity in pair.py

Commit:
updated peek example notebook

Commit:
add n_steps as parameter when running optimizer

jennyfothergill committed:
Standardize docstrings

jennyfothergill committed:
Reformatting

jennyfothergill committed:
Fix r_switch assertion

jennyfothergill committed:
Update string formatting

jennyfothergill committed:
Converted "final.hoomdxml" reference file to gsd

```
import hoomd
from hoomd.deprecated.init import read_xml

sim = hoomd.context.initialize("")
with sim:
    read_xml("msibi/utils/reference/final.hoomdxml")
    hoomd.dump.gsd("final.gsd", period=None, group=hoomd.group.all(), dynamic=['momentum'], phase=-1)
```

jennyfothergill committed:
Saved traj and topology info into query.gsds

```
import mdtraj as md
state_number = 1
topology_filename = "../../utils/reference/final.hoomdxml"
traj_filename = f"../../utils/reference/state{state_number}/query.dcd"
t = md.load(traj_filename, top=topology_filename)
t.save(f"query{state_number}.gsd")
```

jennyfothergill committed:
Copy over reference rdf files

jennyfothergill committed:
Simplify test assets finding

jennyfothergill committed:
Allow state_dir to be passed as kwarg

Update string formatting

jennyfothergill committed:
Update to new schema

MSIBI instance as main container object

jennyfothergill committed:
Remove unused import

jennyfothergill committed:
Scale box/positions by 10

mdtraj was scaling them by /10...

jennyfothergill committed:
Separate state and pair objects

jennyfothergill committed:
Update test schema

jennyfothergill committed:
Allow potential dir to be set

jennyfothergill committed:
Set potential dir to tmp_path

jennyfothergill committed:
Fix logic error--> auto set the path if it _is_ None

jennyfothergill committed:
Create user provided dir if it does not exist

jennyfothergill committed:
Add mkdir command for user provided directory

jennyfothergill committed:
Assume if dir is given user can create it

jennyfothergill committed:
Update imports

jennyfothergill committed:
Update schema

-running optimize(0) computes target rdf
-will need to regenerate rdf assets because freud and mdtraj are
different

jennyfothergill committed:
Remove unused gsd

jennyfothergill committed:
Update docstring

jennyfothergill committed:
Update rdf assets, use smooth and excluded_bonded in test

jennyfothergill committed:
Add alpha param to state

jennyfothergill committed:
Update tests (WIP)

jennyfothergill committed:
Update tests (test_pair working)

jennyfothergill committed:
Remove unused function reload_query_trajectory

jennyfothergill committed:
Update tests

jennyfothergill committed:
Remove unused import and function

jennyfothergill committed:
Remove unused modules/imports

jennyfothergill committed:
Remove unused import

chrisjonesBSU commented:
Looks like all the tests are passing, and this is ready to merge. We can make a separate PR to increase coverage and finish cleaning up the code.
